### PR TITLE
Protect default branch against deletion

### PR DESCRIPTION
**Pull request by 6543**
Although default branch is not offered for deletion in the templates, we need to prevent it both at the router level and in the pre-receive hook.

### CODE REVIEW

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -1683,6 +1683,7 @@ branch.deleted_by = Deleted by %s
 branch.restore_success = Branch '%s' has been restored.
 branch.restore_failed = Failed to restore branch '%s'.
 branch.protected_deletion_failed = Branch '%s' is protected. It cannot be deleted.
+branch.dfault_deletion_failed = Branch '%s' is default branch. It cannot be deleted.
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
branch.default_deletion_failed = Branch '%s' is default branch. It cannot be deleted.
```

File: routers/repo/branch.go [hunk 2]
```diff
@@ -57,8 +57,13 @@ func Branches(ctx *context.Context) {
 // DeleteBranchPost responses for delete merged branch
 func DeleteBranchPost(ctx *context.Context) {
 	defer redirect(ctx)
-
 	branchName := ctx.Query("name")
+
+	if branchName == ctx.Repo.Repository.DefaultBranch {
+		ctx.Flash.Error(ctx.Tr("repo.branch.dfault_deletion_failed", branchName))
```
Comment (zeripath) [inline routers/repo/branch.go]:
```suggestion
		ctx.Flash.Error(ctx.Tr("repo.branch.default_deletion_failed", branchName))
```

**Comment by zeripath** [comment]
Also need to add to pre-receive hook

**Comment by zeripath** [comment]
```diff
--- a/routers/private/hook.go
+++ b/routers/private/hook.go
@@ -206,6 +206,14 @@ func HookPreReceive(ctx *macaron.Context, opts private.Hook
                refFullName := opts.RefFullNames[i]
 
                branchName := strings.TrimPrefix(refFullName, git.BranchPrefix)
+               if branchName == repo.DefaultBranch && newCommitID == git.EmptyS
+                       log.Warn("Forbidden: Branch: %s in %-v is the default br
+                       ctx.JSON(http.StatusForbidden, map[string]interface{}{
+                               "err": fmt.Sprintf("branch %s is the default bra
+                       })
+                       return
+               }
+
                protectBranch, err := models.GetProtectedBranchBy(repo.ID, branc
                if err != nil {
                        log.Error("Unable to get protected branch: %s in %-v Err
```

File: options/locale/locale_en-US.ini (hunk 1 above)
Comment (6543) [reply options/locale/locale_en-US.ini]: haha sure this is wrong

File: routers/repo/branch.go [hunk 3]
```diff
@@ -57,8 +57,13 @@ func Branches(ctx *context.Context) {
 // DeleteBranchPost responses for delete merged branch
 func DeleteBranchPost(ctx *context.Context) {
 	defer redirect(ctx)
-
 	branchName := ctx.Query("name")
+
```
Comment (lafriks) [inline routers/repo/branch.go]: no need for newline here
```suggestion
```

**Review by lafriks — APPROVED** [review]
Apart from small nit, looks good

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -1683,6 +1683,7 @@ branch.deleted_by = Deleted by %s
 branch.restore_success = Branch '%s' has been restored.
 branch.restore_failed = Failed to restore branch '%s'.
 branch.protected_deletion_failed = Branch '%s' is protected. It cannot be deleted.
+branch.default_deletion_failed = Branch '%s' is default branch. It cannot be deleted.
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
branch.default_deletion_failed = Branch '%s' is the default branch. It cannot be deleted.
```

**Review by kolaente — APPROVED** [review]
Looks good.

**Pull request merged**